helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -1045,4 +1048,54 @@ private static void setNextJobPurgeTime(String workflow, long currentTime, long
       rebalanceScheduler.scheduleRebalance(manager, workflow, nextPurgeTime);
     }
   }
+
+  /**
+   * Get target thread pool size from InstanceConfig first; if that fails, get it from
+   * ClusterConfig; if that fails, fall back to the default value.
+   * @param configAccessor - accessor used for the configs
+   * @param clusterName - the cluster name for InstanceConfig and ClusterConfig
+   * @param instanceName - the instance name for InstanceConfig
+   * @return target thread pool size
+   */
+  public static int getTargetThreadPoolSize(ConfigAccessor configAccessor, String clusterName,
+      String instanceName) {
+    // Check instance config first for thread pool size
+    try {
+      InstanceConfig instanceConfig = configAccessor.getInstanceConfig(clusterName, instanceName);
+      if (instanceConfig != null) {
+        int targetTaskThreadPoolSize = instanceConfig.getTargetTaskThreadPoolSize();
+        if (targetTaskThreadPoolSize > 0) {
+          return targetTaskThreadPoolSize;
+        }
+      } else {
+        LOG.warn(
+            "Got null as InstanceConfig for instance {} in cluster {}. Continuing with ClusterConfig. ",
+            instanceName, clusterName);
+      }
+    } catch (HelixException e) {
+      LOG.warn(
+          "Encountered an exception while fetching InstanceConfig for instance {} in cluster {}. Continuing with ClusterConfig. Exception: ",
+          instanceName, clusterName, e);
+    }
+
+    // Fallback to cluster config since instance config doesn't provide the value
+    try {
+      ClusterConfig clusterConfig = configAccessor.getClusterConfig(clusterName);
+      if (clusterConfig != null) {
+        int globalTargetTaskThreadPoolSize = clusterConfig.getGlobalTargetTaskThreadPoolSize();
+        if (globalTargetTaskThreadPoolSize > 0) {
+          return globalTargetTaskThreadPoolSize;
+        }
+      } else {
+        LOG.warn("Got null as ClusterConfig for cluster {}. Returning default value. ",
+            clusterName);
+      }
+    } catch (HelixException e) {
+      LOG.warn(
+          "Encountered an exception while fetching ClusterConfig in cluster {}. Returning default value. Exception: ",
+          clusterName, e);
+    }
+
+    return TaskConstants.DEFAULT_TASK_THREAD_POOL_SIZE;
```

Comment:
If it is HelixException, is it safe to just return the default value?

Reply:
When HelixException is thrown on this line, it means the cluster is not set up. I don't have a strong preference on "returning default" vs "rethrow the exception". What do you think?

Reply:
I would prefer exception because it fails the logic earlier.

Reply:
Resolved offline. Now the logic rethrows exceptions.